scrapy_poet/overrides.py

```diff
@@ -89,35 +70,64 @@ class OverridesRegistry(OverridesRegistryBase):
+        super().__init__(rules=rules)
+        self.overrides_matcher: Dict[Type[ItemPage], URLMatcher] = defaultdict(
+            URLMatcher
         )
+        self.item_matcher: Dict[Optional[Type], URLMatcher] = defaultdict(URLMatcher)
+        for rule_id, rule in enumerate(self._rules):
+            self.add_rule(rule_id, rule)
+        logger.debug(f"List of parsed ApplyRules:\n{self._rules}")
+
+    def add_rule(self, rule_id: int, rule: ApplyRule) -> None:
+        # A common case when a page object subclasses another one with the same
+        # URL pattern. See the ``test_item_return_subclass()`` test case in
+        # ``tests/test_web_poet_rules.py``.
+        matched = self.item_matcher[rule.to_return]
+        pattern_dupes = [
+            pattern
+            for pattern in matched.patterns.values()
+            if pattern == rule.for_patterns
+        ]
+        if pattern_dupes:
+            rules = [
+                r
+                for p in pattern_dupes
+                for r in self.search(for_patterns=p, to_return=rule.to_return)
+            ]
+            warn(
+                f"Similar URL patterns {pattern_dupes} were declared earlier "
+                f"that use to_return={rule.to_return}. The first, highest-priority "
+                f"rule added to SCRAPY_POET_REGISTRY will be used when matching "
+                f"against URLs. Consider updating the priority of these rules: "
+                f"{rules}."
+            )
 
-    def overrides_for(self, request: Request) -> Mapping[Callable, Callable]:
-        overrides: Dict[Callable, Callable] = {}
-        for instead_of, matcher in self.matcher.items():
-            rule_id = matcher.match(request.url)
+        if rule.instead_of:
+            self.overrides_matcher[rule.instead_of].add_or_update(
+                rule_id, rule.for_patterns
+            )
+        if rule.to_return:
+            self.item_matcher[rule.to_return].add_or_update(rule_id, rule.for_patterns)
+
+    # TODO: These URL matching functionalities could be moved to web-poet.
+    # ``overrides_for`` in web-poet could be ``str`` or ``_Url`` subclass.
+
+    def rules_for_url(
+        self, url: str, url_matcher: Dict[Any, URLMatcher]
+    ) -> Mapping[Callable, Callable]:
+        result: Dict[Callable, Callable] = {}
+        for target, matcher in url_matcher.items():
+            rule_id = matcher.match(url)
             if rule_id is not None:
-                overrides[instead_of] = self.rules[rule_id].use
-        return overrides
+                result[target] = self._rules[rule_id].use
+        return result
+
+    def overrides_for(self, request: Request) -> Mapping[Callable, Callable]:
+        return self.rules_for_url(request.url, self.overrides_matcher)
+
+    def page_object_for_item(
+        self, url: str, po_cls: Type[Any]
+    ) -> Optional[Callable[..., Any]]:
+        return self.rules_for_url(url, self.item_matcher).get(po_cls)
```

Comment:
Can it be optimized? Currently it runs matcher.match for everything in the registry; it seems running it once, for the item_cls, should be enough.

Reply:
Updated how it's implemented in https://github.com/scrapinghub/web-poet/pull/112 wherein its now returning a `Mapping` and doesn't accept an `item_cls` anymore.